File: web/qz-tray.ts
type Nullable<T> = T | null;

interface PrintPayload {
    printer: string,
    config: any,
    data: any
}

class QZTray {
    private _qz: any;

    constructor() {
        this._qz = window['qz'];
    }

    setCertificate(certificate: string): void {
        this._qz.security.setCertificatePromise((resolve) => {
            return resolve(certificate);
        });
    }

    setSignature(sign: (toSign: string) => string): void {
        this._qz.security.setSignaturePromise((_: string) => {
            return (resolve, reject) => {
                return resolve(sign(_));
            }
        });
    }

    connect(): Promise<Operation | void> {
        return new Promise<Operation | void>((resolve, reject) => {
            if (this._qz.websocket.isActive()) {
                resolve();
            }

            this._qz.websocket
            .connect()
            .then(() => {
                resolve();
            })
            .catch((err: any) => {
                reject(new Operation(null, err));
            })
        });
    }
    
    getPrinters(): Promise<Operation> {
        return new Promise<Operation>((resolve, reject) => {
            this.connect()
            .then(() => {
                this._qz.printers
                .find()
                .then((printers: string | string[]) => {
                    resolve(new Operation(printers, null));
                })
                .catch((err: any) => {
                    reject(new Operation(null, err));
                });
            })
            .catch((errConn: Operation) => {
                reject(new Operation(null, errConn.error));
            });

        });
    }

    print(payload: PrintPayload): Promise<Operation | void> {
        return new Promise<Operation | void>((resolve, reject) => {
            let qzTrayConfig: any
            try {
                qzTrayConfig = this._qz.configs.create(payload.printer, payload.config);
            } catch(err: any) {
                reject(reject(new Operation(null, err)));
            }
            
            this.connect()
            .then(() => {
                this._qz.print(qzTrayConfig, payload.data)
                .then(() => {
                    resolve();
                })
                .catch((err: any) => {
                    reject(new Operation(null, err))
                });
            })
            .catch((errConn: Operation) => {
                reject(new Operation(null, errConn.error));
            });
        });
    }
}

class Operation {
    private _data: any;
    private _error: Nullable<string>;

    constructor(data: any, error: Nullable<string>) {
        this._data = data;
        this._error = error;
    }

    get data(): any {
        return this._data;
    }

    get error(): any {
        return this._error;
    }
}

File: lib/qz_tray/qz_tray.dart
@JS()
library qz_tray;

import 'dart:developer';

import 'package:js/js.dart';
import 'package:js/js_util.dart';
import 'package:qz_tray/exception/connection_exception.dart';
import 'package:qz_tray/exception/find_printers_exception.dart';
import 'package:qz_tray/exception/print_exception.dart';

class QZTray {
  final interface = _QZTray();

  void setCertificate() {
    interface.setCertificate('Certificate11');
  }

  void setSignature() {
    interface.setSignature(allowInterop((toSign) {
      log('=====To sign====');
      log(toSign);
      return 'signature perro';
    }));
  }

  Future<void> connect() async {
    try {
      await promiseToFuture(interface.connect());
    } on Operation catch(operation) {
      throw ConnectionException(operation.error ?? 'Error trying to connect to qz tray driver.');
    }
  }

  Future<Operation> getPrinters() async {
    try {
      return await promiseToFuture<Operation>(interface.getPrinters());
    } on Operation catch(operation) {
      throw FindPrintersException(operation.error ?? 'Error trying to find printers.');
    }
  }

  Future<void> print(PrintPayload payload) async {
    try {
      await promiseToFuture<Operation>(interface.print(payload));
    } on Operation catch(operation) {
      throw PrintException(operation.error ?? 'Error trying to print.');
    }
  }
}

@JS('QZTray')
class _QZTray {
  external _QZTray();
  external void setCertificate(String certificate);
  external void setSignature(String Function(String toSign) func);
  external dynamic connect();
  external dynamic getPrinters();
  external dynamic print(PrintPayload payload);
}

@JS()
class Operation {
  external dynamic get data;
  external String? get error;
  external factory Operation(int status, dynamic data, String? error);
}

@JS()
@anonymous
class PrintPayload {
  external String get printer;
  external dynamic get config;
  external dynamic get data;
  external factory PrintPayload({String printer, dynamic config, dynamic data});
}